yarn-project/end-to-end/src/e2e_fees.test.ts

```diff
@@ -312,3 +423,81 @@ describe('e2e_fees', () => {
     await e2eContext.wallets[accountIndex].addNote(extendedNote);
   };
 });
+
+class BuggedSetupFeePaymentMethod extends PublicFeePaymentMethod {
+  getFunctionCalls(maxFee: Fr): Promise<FunctionCall[]> {
+    const nonce = Fr.random();
+    const messageHash = computeAuthWitMessageHash(this.paymentContract, {
+      args: [this.wallet.getAddress(), this.paymentContract, maxFee, nonce],
+      functionData: new FunctionData(
+        FunctionSelector.fromSignature('transfer_public((Field),(Field),Field,Field)'),
+        false,
+        false,
+        false,
+      ),
+      to: this.asset,
+    });
+
+    const tooMuchFee = new Fr(maxFee.toBigInt() * 2n);
+
+    return Promise.resolve([
+      this.wallet.setPublicAuth(messageHash, true).request(),
+      {
+        to: this.getPaymentContract(),
+        functionData: new FunctionData(
+          FunctionSelector.fromSignature('fee_entrypoint_public(Field,(Field),Field)'),
+          false,
+          true,
+          false,
+        ),
+        args: [tooMuchFee, this.asset, nonce],
+      },
+    ]);
+  }
+}
+
+class BuggedTeardownFeePaymentMethod extends PublicFeePaymentMethod {
+  async getFunctionCalls(maxFee: Fr): Promise<FunctionCall[]> {
+    // authorize the FPC to take the max fee from Alice
+    const nonce = Fr.random();
+    const messageHash1 = computeAuthWitMessageHash(this.paymentContract, {
+      args: [this.wallet.getAddress(), this.paymentContract, maxFee, nonce],
+      functionData: new FunctionData(
+        FunctionSelector.fromSignature('transfer_public((Field),(Field),Field,Field)'),
+        false,
+        false,
+        false,
+      ),
+      to: this.asset,
+    });
+
+    // authorize the FPC to take the maxFee
+    // do this first because we only get 2 feepayload calls
+    await this.wallet.setPublicAuth(messageHash1, true).send().wait();
```

Comment:
I think a different failure scenario that doesn't involve frontrunning the tx with a nother tx would be for the bugged payment method to transfer `MaxFee/2` tokens from Alice causing the FPC to run out of public funds and not be able to process the refund (similar to the bugged setup above that tries to take `MaxFee*2`)